### DISCUSSION TITLE
Allow to export output of function as something else appart from $input

### DISCUSSION THREAD

**Discussion by ianchi** [microsoft/semantic-kernel#1540] [Ideas]
### Current Situtation
There are many scenarios in which allowing a function (semantic or native) to export the output result as something else (in addition to) `$input` would be useful:

+ the next function in the chain needs the input in another/additional variable. Now these variables can only be manually set in the context from outside the chain.
+ you need to keep an intermediate result of the chain for further processing. But `$input` gets overwritten with the result of the following chain, and so it is lost.
+ combine outputs from two/more previous steps in different parts of the prompt for a following semantic function.

All these scenarios cannot be accomplished now using a single chain of functions. They all force to break the chain into smaller blocks, and then manually manipulate the context for the next step(s) and run a new subchain.

This completely ruins a very elegant piping design.

### Proposed idea

+ For _Semantic Functions_ add an `output` property in the `config.json` to specify an additional variable where to store the output (in addition to `$input`)
+ For _Native Functions_ add an `output` parameter in the decorator, with the same logic

If you agree with the idea, I could send a PR implementing it for the Python version.

**Comment by ianchi** [comment]
As a workarround I built a factory for volatile SKFunction to put in the middle of the chain and copy to new variable:

#### Helper
```python
def copy_variable(destination: str, source: str = "input") -> SKFunction:
    def helper_function(input_: str, context: SKContext) -> str:
        context.variables[destination] = context.variables[source]
        return input_

    return SKFunction(
        delegate_type=DelegateInference.infer_delegate_type(helper_function),
        delegate_function=helper_function,
        parameters=[ParameterView("input", "Input", "")],
        description="Copy source context variable to destination",
        skill_name="",
        function_name="copy_variable",
        is_semantic=False,
    )
```

#### Usage:
```python
result = await kernel.run_async(
       semantic_functions["extractFunction"],
       copy_variable("entities"),
       native_function["get_sources"],
       copy_variable("sources"),
       semantic_functions["answer"],
       input_vars=context,
)

# get intermediate values
sources = result.variables["sources"]
entities = result.variables["entities"]
```

**Comment by moonbox3** [comment]
Hi @ianchi, we've made a ton of updates to the Python code since this question was asked. For example, in our latest beta package, we return a `FunctionResult` object instead of an `SKFunction`. Can you please go through the latest code/kernel examples/samples, etc and let us know if your question still stands. I am closing this since the related code no longer exists. Please feel free to start a new discussion as needed. Thanks.
